### ISSUE TITLE
SubsamplingLayer Null Pointer Exception on toFlattened (CpuNDArrayFactory)

### ISSUE THREAD

**Issue by ShakMR** [deeplearning4j/deeplearning4j#1358]
When using the master (version 0.4-rc3.9) branch of DeepLearning4j and I'm getting some Null Pointer.
The network definition is this (copied from the repository [LenetMnistExample](https://github.com/deeplearning4j/dl4j-0.4-examples/blob/master/src/main/java/org/deeplearning4j/examples/convolution/LenetMnistExample.java)):

        builder = new NeuralNetConfiguration.Builder()
                .seed(seed)
                .iterations(iterations)
                .regularization(true).l2(0.0005)
                .learningRate(0.01)
                .weightInit(WeightInit.XAVIER)
                .optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT)
                .updater(Updater.NESTEROVS).momentum(0.9)
                .list(4)
                .layer(0, new ConvolutionLayer.Builder(5, 5)
                        .nIn(nChannels)
                        .stride(1, 1)
                        .nOut(20).dropOut(0.5)
                        .activation("relu")
                        .build())
                .layer(1, new SubsamplingLayer.Builder(SubsamplingLayer.PoolingType.MAX)
                        .kernelSize(2,2)
                        .stride(2,2)
                        .build())
                .layer(2, new DenseLayer.Builder().activation("relu")
                        .nOut(500).build())
                .layer(3, new OutputLayer.Builder(LossFunctions.LossFunction.NEGATIVELOGLIKELIHOOD)
                        .nOut(outputNum)
                        .activation("softmax")
                        .build())
                .backprop(true).pretrain(false);

The exception showned is this:

    Exception in thread "main" java.lang.NullPointerException
    	at org.nd4j.linalg.cpu.nativecpu.CpuNDArrayFactory.toFlattened(CpuNDArrayFactory.java:489)
    	at org.nd4j.linalg.factory.Nd4j.toFlattened(Nd4j.java:1575)
    	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.params(MultiLayerNetwork.java:861)
    	at org.deeplearning4j.spark.impl.multilayer.SparkDl4jMultiLayer.runIteration(SparkDl4jMultiLayer.java:374)
    	at org.deeplearning4j.spark.impl.multilayer.SparkDl4jMultiLayer.fitDataSet(SparkDl4jMultiLayer.java:344)
    	at org.deeplearning4j.ml.JavaMNistClass.main(JavaMNistClass.java:285)
    	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    	at java.lang.reflect.Method.invoke(Method.java:498)
    	at org.apache.spark.deploy.SparkSubmit$.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:672)
    	at org.apache.spark.deploy.SparkSubmit$.doRunMain$1(SparkSubmit.scala:180)
    	at org.apache.spark.deploy.SparkSubmit$.submit(SparkSubmit.scala:205)
    	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:120)
    	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)

I've debbuged this using an IDE and find that the *NullPointerException* is thrown due to a null element in a *Collection<INDArray>* in:

CpuNDArrayFactory.java:489

Researching for the origin of this null element, I've found that it is originated when (in *org.deeplearning4j.nn.multilayer.MultiLayerNetwork.params()* around line 858) the layer number 1 (which correspons to the *SubsamplingLayer*) didn't have any parameter.

Indeed, you cannot set any parameter to the *SubsamplingLayer*, taking in account that this parameters are the *nIn* and *nOut* functions of the other layers (guessed by debugging).

**Comment by AlexDBlack** [comment]
Does this occur on the LenetMnistExample or any other (unmodified) example?
I ask because I've run pretty much all of the examples recently (on 3.9) and I'm not seeing this problem on any of them.

**Comment by ShakMR** [comment]
It is a modified example of the LenetMnistExample. The only difference is that my example reads data from the local file-system with a function like:

    public static Iterator<DataSet> readMnistFile(Uri imgPath, Uri lblPath, int records) throws Exception

I will test the unmodified LenetMnistExample on the system as soon as I can. 
In rc3.8 I haven't found this problem.

**Comment by ShakMR** [comment]
Well, I've executed the LenetMnistExample without problem. So I've deduced that the problem arises when using Spark.

I've seen that the example (without Spark) goes through this params function 

    /**
     * Returns a 1 x m vector where the vector is composed of
     * a flattened vector of all of the weights for the
     * various neuralNets(w,hbias NOT VBIAS) and output layer
     *
     * @return the params for this neural net
     */
    @Override
    public INDArray params() {
        if(params != null)
            return params;

        List<INDArray> params = new ArrayList<>();
        for (Layer layer: getLayers()){
            INDArray layerParams;
            if( layer instanceof BasePretrainNetwork){
                layerParams = ((BasePretrainNetwork) layer).paramsBackprop();
            }
            else layerParams = layer.params();
            if(layerParams != null) params.add(layerParams);    //may be null: subsampling etc layers
        }
        return Nd4j.toFlattened('f',params);
    }

While in my example (with Spark) goes throught this other

    /**
     * Returns a 1 x m vector where the vector is composed of
     * a flattened vector of all of the weights for the
     * various neuralNets(w,hbias NOT VBIAS) and output layer
     *
     * @return the params for this neural net
     */
    public INDArray params(boolean backwardOnly) {
        if(params != null)
            return params;

        List<INDArray> params = new ArrayList<>();
        for (Layer layer: getLayers()){
            if( layer instanceof BasePretrainNetwork && backwardOnly)
                params.add(((BasePretrainNetwork) layer).paramsBackprop());
            else
                params.add(layer.params());
        }

        return Nd4j.toFlattened('f', params);
    }

The first one is taking into account that some layers cat have a *layerParams*.
I copied that "Null check" to the other function as well. And, as a result the training task began to execute (which wasn't happenning before)
Unfontunetly a bigger problem appears (bigger in the error output at least).

    16/04/05 01:00:55 WARN TaskSetManager: Lost task 0.0 in stage 0.0 (TID 0, 192.168.1.10): java.lang.IllegalStateException: Unable to get number of of columns for a non 2d matrix
	at org.nd4j.linalg.api.ndarray.BaseNDArray.columns(BaseNDArray.java:3424)
	at org.nd4j.linalg.dataset.DataSet.merge(DataSet.java:117)
	at org.deeplearning4j.spark.impl.multilayer.IterativeReduceFlatMap.call(IterativeReduceFlatMap.java:84)
	at org.deeplearning4j.spark.impl.multilayer.IterativeReduceFlatMap.call(IterativeReduceFlatMap.java:48)
	at org.apache.spark.api.java.JavaRDDLike$$anonfun$fn$5$1.apply(JavaRDDLike.scala:167)
	at org.apache.spark.api.java.JavaRDDLike$$anonfun$fn$5$1.apply(JavaRDDLike.scala:167)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitions$1$$anonfun$apply$17.apply(RDD.scala:706)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitions$1$$anonfun$apply$17.apply(RDD.scala:706)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:297)
	at org.apache.spark.CacheManager.getOrCompute(CacheManager.scala:69)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:262)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:297)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:264)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:66)
	at org.apache.spark.scheduler.Task.run(Task.scala:88)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:214)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

**Comment by nyghtowl** [comment]
@ShakMR thanks for catching that
 #1373 should address the problem
Let us know if it doesn't
